internal/provider/connection_string_data_source.go

```diff
@@ -0,0 +1,172 @@
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Required:    true,
+				Description: "Cluster ID",
+			},
+			"os": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				Description: "Used to determine the SSL certificate path for dedicated clusters. " +
+					"Defaults to the Terraform user's OS. " +
+					"Options are 'WINDOWS', 'MAC', and 'LINUX'.",
+			},
+			"database": schema.StringAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: fmt.Sprintf("Database to connect to. Defaults to '%s'.", defaultDB),
+			},
+			"sql_user": schema.StringAttribute{
+				Optional:    true,
+				Description: "Database username.",
+			},
+			"password": schema.StringAttribute{
+				Optional:            true,
+				Sensitive:           true,
+				MarkdownDescription: "Database user password. Must also include `sql_user`.",
+			},
+			"connection_string": schema.StringAttribute{
+				Computed:    true,
+				Description: "Fully formatted connection string. Assumes the cluster certificate is stored in the default location.",
+			},
+			"connection_params": schema.MapAttribute{
+				Computed:    true,
+				ElementType: types.StringType,
+				Description: "List of individual connection string parameters. Can be used to build nonstandard connection strings.",
+			},
+		},
+		Description: "Generic connection string for a given cluster",
+	}
+}
+
+func (d *connectionStringDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_connection_string"
+}
+
+func (d *connectionStringDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	var ok bool
+	if d.provider, ok = req.ProviderData.(*provider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", provider{}, req.ProviderData))
+	}
+}
+
+func (d *connectionStringDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	if d.provider == nil || !d.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var config ConnectionString
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		resp.Diagnostics.AddWarning("Error loading the connection string config", "")
+		return
+	}
+
+	database := defaultDB
+	if config.Database.IsNull() {
+		config.Database = types.StringValue(database)
+	} else {
+		database = config.Database.ValueString()
+	}
+	var sqlUser *string
+	if !config.SqlUser.IsNull() {
+		sqlUser = new(string)
+		*sqlUser = config.SqlUser.ValueString()
+	}
+	var os string
+	if !config.OS.IsNull() {
+		os = fmt.Sprintf("OS_%s", config.OS.ValueString())
+	} else {
+		switch runtime.GOOS {
+		case "windows":
+			os = "OS_WINDOWS"
```

Comment:
Maybe I missed this elsewhere, it seems weird to me to prefix the string values with "OS_".   thoughts?